### ISSUE TITLE
Parallel computation

### ISSUE THREAD

**Issue by freakingwave** [mancellin/capytaine#65]
I have a 64-core workstation at hand. When I run Capytaine it only takes less than 10% of CPU. 
How can I deploy all CPU cores to run a task?

When I run it on a 8-core PC, it automatically takes 100% of the CPU.

Thanks

**Comment by mancellin** [comment]
The number of threads is controlled by the environment variables `OMP_NUM_THREADS` and `MKL_NUM_THREADS`. Could you check their values?

**Comment by freakingwave** [comment]
> 
> 
> The number of threads is controlled by the environment variables `OMP_NUM_THREADS` and `MKL_NUM_THREADS`. Could you check their values?

Thank you for reply.

by setting:
import os
 os.environ['OMP_NUM_THREADS']='32'
 os.environ['MKL_NUM_THREADS']='32'
 os.environ['NUMEXPR_MAX_THREADS'] = '32'

It still only takes less than 10% CPU. Time costs have no difference with setting 1, 8, 16, respectively.

Can you give some insights?

**Comment by mancellin** [comment]
Changing the value from within Python seems to be buggy (see also #47).

If it is not the issue, could you tell me more about your OS and the version of Capytaine?

**Comment by freakingwave** [comment]
I see...
I am using the latest 1.2.1 Capytaine by Spyder on Windows 10.

**Comment by mancellin** [comment]
You should look on the internet for a way to change these variables system-wide on Windows 10 (sorry I can't help you with that).

Could you also check if it behaves the same with the previous version? (`conda install -c conda-forge capytaine=1.2.0`)

**Comment by freakingwave** [comment]
Thanks.

1. Capy 1.2.0 behaves the same.
2. I set the global variables on Win10. It is weird. When I set threads to 1, it is obvious that only one thread is working. When I set them to a large number of threads like 16, 32 or 64, the CPU useage (about 12~15%) and cost time are the same.

**Comment by mancellin** [comment]
I suppose it is not a bug, just the usual problem of diminishing returns in HPC...
I don't remember myself testing Capytaine with more than 8 cores. Maybe adding more threads does not improve the efficiency after some point. Or maybe your test problem is too simple to be worth using all the cores (the parallelism is on the number of panels in the mesh).

You could still maybe use all of your available power by setting OMP_NUM_THREADS to 4 or 8 and by launching at the same time several python scripts with different ranges of frequencies (for instance). But, you'll have to do it manually, there is no automated tools for that at the moment.

**Comment by freakingwave** [comment]
Thank you for your suggestions.

I agree with you about the idea adding more threads will not obviously imrove the efficiency. However, I think the CPU usage will show some difference when adding more threads no matter what efficiency it shows.

I will try other ways. Thanks again.

**Comment by ryancoe** [comment]
@freakingwave - I have noticed something similar on macOS, but don't yet understand the cause. Here's my scenario:

I have an environment with Capytaine 1.20 (see below). I can clearly see that the examples in Capytaine (e.g., [`radiation_cylinder.py`](https://github.com/mancellin/capytaine/blob/87d09bf56dcd1d4a242a29805e9d49511b3947e1/docs/user_manual/examples/radiation_cylinder.py)) use all my cores. However, I'm working on a tool that incorporates Capytaine into a larger work flow for doing design optimization. As such, I've wrapped Capytaine inside another class, etc. Something about this seems to prevent it from running in parallel. I've done only a little bit of experimenting, but I haven't been able to figure out what the issue is yet. @mancellin - Can you point me to something showing/explaining what tools Capytaine is using for parallel computations (this isn't yet clear to me from the source)?

<details><summary>env listing</summary>
<p>

```
$ conda list
# packages in environment at /Users/rcoe/anaconda3/envs/_wecopttool:
#
# Name                    Version                   Build  Channel
absl-py                   0.12.0             pyhd8ed1ab_0    conda-forge
alabaster                 0.7.12                     py_0    conda-forge
ampl-mp                   3.1.0             h1423387_1004    conda-forge
anyio                     2.2.0            py39h6e9494a_0    conda-forge
appdirs                   1.4.4              pyh9f0ad1d_0    conda-forge
applaunchservices         0.2.1                      py_0    conda-forge
appnope                   0.1.2            py39h6e9494a_1    conda-forge
argcomplete               1.12.2             pyhd8ed1ab_0    conda-forge
argh                      0.26.2          pyh9f0ad1d_1002    conda-forge
argon2-cffi               20.1.0           py39hcbf5805_2    conda-forge
astroid                   2.5.3            py39h6e9494a_0    conda-forge
async_generator           1.10                       py_0    conda-forge
atomicwrites              1.4.0              pyh9f0ad1d_0    conda-forge
attrs                     20.3.0             pyhd3deb0d_0    conda-forge
autograd                  1.3                        py_0    conda-forge
autopep8                  1.5.6              pyhd8ed1ab_0    conda-forge
babel                     2.9.0              pyhd3deb0d_0    conda-forge
backcall                  0.2.0              pyh9f0ad1d_0    conda-forge
backports                 1.0                        py_2    conda-forge
backports.functools_lru_cache 1.6.4              pyhd8ed1ab_0    conda-forge
black                     20.8b1                     py_1    conda-forge
bleach                    3.3.0              pyh44b312d_0    conda-forge
brotlipy                  0.7.0           py39hcbf5805_1001    conda-forge
bzip2                     1.0.8                hc929b4f_4    conda-forge
c-ares                    1.17.1               h0d85af4_1    conda-forge
ca-certificates           2020.12.5            h033912b_0    conda-forge
cached-property           1.5.2                hd8ed1ab_1    conda-forge
cached_property           1.5.2              pyha770c72_1    conda-forge
capytaine                 1.2.1            py39h70c5cf5_0    conda-forge
casadi                    3.5.5            py39hcc51aa2_3    conda-forge
certifi                   2020.12.5        py39h6e9494a_1    conda-forge
cffi                      1.14.5           py39h319c39b_0    conda-forge
cftime                    1.4.1            py39h329c335_0    conda-forge
chardet                   4.0.0            py39h6e9494a_1    conda-forge
click                     7.1.2              pyh9f0ad1d_0    conda-forge
cloudpickle               1.6.0                      py_0    conda-forge
colorama                  0.4.4              pyh9f0ad1d_0    conda-forge
control                   0.9.0            py39h6e9494a_0    conda-forge
cryptography              3.4.7            py39ha2c9959_0    conda-forge
curl                      7.76.0               h06286d4_0    conda-forge
cvxpy                     1.1.11           py39h6e9494a_0    conda-forge
cvxpy-base                1.1.11           py39h4d6be9b_0    conda-forge
cycler                    0.10.0                     py_2    conda-forge
cyipopt                   1.0.3            py39h5325327_0    conda-forge
dataclasses               0.8                pyhc8e2a94_1    conda-forge
dbus                      1.13.6               ha13b53f_2    conda-forge
decorator                 5.0.6              pyhd8ed1ab_0    conda-forge
defusedxml                0.7.1              pyhd8ed1ab_0    conda-forge
diff-match-patch          20200713           pyh9f0ad1d_0    conda-forge
docutils                  0.17             py39h6e9494a_0    conda-forge
double-conversion         3.1.5                he49afe7_2    conda-forge
ecos                      2.0.8            py39hc89836e_0    conda-forge
eigen                     3.3.9                h926bf3e_1    conda-forge
entrypoints               0.3             pyhd8ed1ab_1003    conda-forge
expat                     2.3.0                he49afe7_0    conda-forge
fatpack                   0.7.0                    pypi_0    pypi
ffmpeg                    4.3.1                hb787a4c_2    conda-forge
flake8                    3.8.4                      py_0    conda-forge
freetype                  2.10.4               h4cff582_1    conda-forge
future                    0.18.2           py39h6e9494a_3    conda-forge
gettext                   0.19.8.1          h7937167_1005    conda-forge
gl2ps                     1.4.2                h4cff582_0    conda-forge
glew                      2.1.0                h046ec9c_2    conda-forge
glib                      2.68.1               he49afe7_0    conda-forge
glib-tools                2.68.1               he49afe7_0    conda-forge
gmp                       6.2.1                h2e338ed_0    conda-forge
gmsh                      4.8.3                    pypi_0    pypi
gnutls                    3.6.13               h756fd2b_1    conda-forge
h5py                      3.2.1           nompi_py39h1bb8402_100    conda-forge
hdf4                      4.2.13            h71d84a9_1004    conda-forge
hdf5                      1.10.6          nompi_hc5d9132_1114    conda-forge
icu                       68.1                 h74dc148_0    conda-forge
idna                      2.10               pyh9f0ad1d_0    conda-forge
imagesize                 1.2.0                      py_0    conda-forge
importlib-metadata        3.10.1           py39h6e9494a_0    conda-forge
importlib_metadata        3.10.1               hd8ed1ab_0    conda-forge
iniconfig                 1.1.1              pyh9f0ad1d_0    conda-forge
intervaltree              3.0.2                      py_0    conda-forge
ipopt                     3.13.4               h8d9dcd1_0    conda-forge
ipykernel                 5.5.3            py39h71a6800_0    conda-forge
ipython                   7.22.0           py39h71a6800_0    conda-forge
ipython_genutils          0.2.0                      py_1    conda-forge
isort                     5.8.0              pyhd8ed1ab_0    conda-forge
jax                       0.2.11             pyhd8ed1ab_0    conda-forge
jaxlib                    0.1.65           py39h6e9494a_0    conda-forge
jedi                      0.17.2           py39h6e9494a_1    conda-forge
jinja2                    2.11.3             pyh44b312d_0    conda-forge
joblib                    1.0.1              pyhd8ed1ab_0    conda-forge
jpeg                      9d                   hbcb3906_0    conda-forge
json5                     0.9.5              pyh9f0ad1d_0    conda-forge
jsoncpp                   1.9.4                h926bf3e_1    conda-forge
jsonschema                3.2.0              pyhd8ed1ab_3    conda-forge
jupyter-packaging         0.7.12             pyhd8ed1ab_0    conda-forge
jupyter_client            6.1.12             pyhd8ed1ab_0    conda-forge
jupyter_core              4.7.1            py39h6e9494a_0    conda-forge
jupyter_server            1.6.1            py39h6e9494a_0    conda-forge
jupyterlab                3.0.10             pyhd8ed1ab_0    conda-forge
jupyterlab_pygments       0.1.2              pyh9f0ad1d_0    conda-forge
jupyterlab_server         2.4.0              pyhd8ed1ab_0    conda-forge
keyring                   23.0.1           py39h6e9494a_0    conda-forge
kiwisolver                1.3.1            py39hedf5dff_1    conda-forge
krb5                      1.17.2               h60d9502_0    conda-forge
lame                      3.100             h35c211d_1001    conda-forge
lazy-object-proxy         1.6.0            py39h89e85a6_0    conda-forge
lcms2                     2.12                 h577c468_0    conda-forge
libblas                   3.9.0                8_openblas    conda-forge
libcblas                  3.9.0                8_openblas    conda-forge
libclang                  11.1.0          default_he082bbe_0    conda-forge
libcurl                   7.76.0               h8ef9fac_0    conda-forge
libcxx                    11.1.0               habf9029_0    conda-forge
libedit                   3.1.20191231         h0678c8f_2    conda-forge
libev                     4.33                 haf1e3a3_1    conda-forge
libffi                    3.3                  h046ec9c_2    conda-forge
libgfortran               5.0.0           9_3_0_h6c81a4c_22    conda-forge
libgfortran5              9.3.0               h6c81a4c_22    conda-forge
libglib                   2.68.1               hd556434_0    conda-forge
libiconv                  1.16                 haf1e3a3_0    conda-forge
liblapack                 3.9.0                8_openblas    conda-forge
libllvm11                 11.1.0               hd011deb_2    conda-forge
libnetcdf                 4.7.4           nompi_h9d8a93f_107    conda-forge
libnghttp2                1.43.0               h07e645a_0    conda-forge
libogg                    1.3.4                h35c211d_1    conda-forge
libopenblas               0.3.12          openmp_h54245bb_1    conda-forge
libpng                    1.6.37               h7cec526_2    conda-forge
libpq                     13.1                 h052a64a_2    conda-forge
libsodium                 1.0.18               hbcb3906_1    conda-forge
libspatialindex           1.9.3                h1c7c35f_3    conda-forge
libssh2                   1.9.0                h52ee1ee_6    conda-forge
libtheora                 1.1.1             haf1e3a3_1004    conda-forge
libtiff                   4.2.0                h355d032_0    conda-forge
libvorbis                 1.3.7                h046ec9c_0    conda-forge
libwebp-base              1.2.0                h0d85af4_2    conda-forge
libxml2                   2.9.10               h93ec3fd_4    conda-forge
libxslt                   1.1.33               h5739fc3_2    conda-forge
llvm-openmp               11.1.0               hda6cdc1_1    conda-forge
loguru                    0.5.3            py39hde42818_2    conda-forge
lxml                      4.6.2                    pypi_0    pypi
lz4-c                     1.9.3                h046ec9c_0    conda-forge
markupsafe                1.1.1            py39hcbf5805_3    conda-forge
matplotlib-base           3.4.1            py39hb07454d_0    conda-forge
mccabe                    0.6.1                      py_1    conda-forge
meshio                    4.3.10                   pypi_0    pypi
meshmagick                2.1.1                    pypi_0    pypi
metis                     5.1.0             h2e338ed_1006    conda-forge
mhkit                     0.3.1                    pypi_0    pypi
mistune                   0.8.4           py39hcbf5805_1003    conda-forge
more-itertools            8.7.0              pyhd8ed1ab_0    conda-forge
mumps-include             5.2.1               h694c41f_10    conda-forge
mumps-seq                 5.2.1               h994990c_10    conda-forge
mypy_extensions           0.4.3            py39h6e9494a_3    conda-forge
mysql-common              8.0.23               h694c41f_1    conda-forge
mysql-libs                8.0.23               hbeb7981_1    conda-forge
nbclassic                 0.2.7              pyhd8ed1ab_0    conda-forge
nbclient                  0.5.3              pyhd8ed1ab_0    conda-forge
nbconvert                 6.0.7            py39h6e9494a_3    conda-forge
nbformat                  5.1.3              pyhd8ed1ab_0    conda-forge
ncurses                   6.2                  h2e338ed_4    conda-forge
nest-asyncio              1.5.1              pyhd8ed1ab_0    conda-forge
netcdf4                   1.5.6           nompi_py39h353b61e_101    conda-forge
nettle                    3.6                  hedd7734_0    conda-forge
notebook                  6.3.0            py39h6e9494a_0    conda-forge
nspr                      4.30                 hcd9eead_0    conda-forge
nss                       3.47                 hc0980d9_0    conda-forge
numpy                     1.20.2           py39h7eed0ac_0    conda-forge
numpydoc                  1.1.0                      py_1    conda-forge
olefile                   0.46               pyh9f0ad1d_1    conda-forge
openh264                  2.1.1                hfd3ada9_0    conda-forge
openjpeg                  2.4.0                h6cbf5cd_0    conda-forge
openssl                   1.1.1k               h0d85af4_0    conda-forge
opt_einsum                3.3.0              pyhd8ed1ab_1    conda-forge
osqp                      0.6.2            py39h4d6be9b_1    conda-forge
packaging                 20.9               pyh44b312d_0    conda-forge
pandas                    1.2.4            py39h4d6be9b_0    conda-forge
pandoc                    2.13                 h0d85af4_0    conda-forge
pandocfilters             1.4.2                      py_1    conda-forge
parso                     0.7.0              pyh9f0ad1d_0    conda-forge
pathspec                  0.8.1              pyhd3deb0d_0    conda-forge
pcre                      8.44                 hb1e8313_0    conda-forge
pecos                     0.2.0                    pypi_0    pypi
pexpect                   4.8.0              pyh9f0ad1d_2    conda-forge
pickleshare               0.7.5                   py_1003    conda-forge
pillow                    8.1.2            py39h5fdd921_1    conda-forge
pip                       21.0.1             pyhd8ed1ab_0    conda-forge
pluggy                    0.13.1           py39h6e9494a_4    conda-forge
proj                      7.2.0                h78d1473_2    conda-forge
prometheus_client         0.10.1             pyhd8ed1ab_0    conda-forge
prompt-toolkit            3.0.18             pyha770c72_0    conda-forge
psutil                    5.8.0            py39hcbf5805_1    conda-forge
ptyprocess                0.7.0              pyhd3deb0d_0    conda-forge
pugixml                   1.11.4               he49afe7_0    conda-forge
py                        1.10.0             pyhd3deb0d_0    conda-forge
pycodestyle               2.6.0              pyh9f0ad1d_0    conda-forge
pycparser                 2.20               pyh9f0ad1d_2    conda-forge
pydocstyle                6.0.0              pyhd8ed1ab_0    conda-forge
pyflakes                  2.2.0              pyh9f0ad1d_0    conda-forge
pygments                  2.8.1              pyhd8ed1ab_0    conda-forge
pygmsh                    7.1.8                    pypi_0    pypi
pylint                    2.7.2            py39h6e9494a_0    conda-forge
pyls-black                0.4.6              pyh9f0ad1d_0    conda-forge
pyls-spyder               0.3.2              pyhd8ed1ab_0    conda-forge
pyopenssl                 20.0.1             pyhd8ed1ab_0    conda-forge
pyparsing                 2.4.7              pyh9f0ad1d_0    conda-forge
pyqt                      5.12.3           py39h6e9494a_7    conda-forge
pyqt-impl                 5.12.3           py39hef7122c_7    conda-forge
pyqt5-sip                 4.19.18          py39hd8f94c5_7    conda-forge
pyqtchart                 5.12             py39hef7122c_7    conda-forge
pyqtwebengine             5.12.1           py39hef7122c_7    conda-forge
pyrsistent                0.17.3           py39hcbf5805_2    conda-forge
pysocks                   1.7.1            py39h6e9494a_3    conda-forge
pytest                    6.2.3            py39h6e9494a_0    conda-forge
pytest-repeat             0.8.0                      py_0    conda-forge
python                    3.9.2           h2502468_0_cpython    conda-forge
python-dateutil           2.8.1                      py_0    conda-forge
python-flatbuffers        1.12               pyhd8ed1ab_1    conda-forge
python-jsonrpc-server     0.4.0              pyh9f0ad1d_0    conda-forge
python-language-server    0.36.2             pyhd8ed1ab_0    conda-forge
python.app                1.3              py39hcbf5805_4    conda-forge
python_abi                3.9                      1_cp39    conda-forge
pytz                      2021.1             pyhd8ed1ab_0    conda-forge
pyyaml                    5.4.1            py39hcbf5805_0    conda-forge
pyzmq                     22.0.3           py39h7fec2f1_1    conda-forge
qdarkstyle                3.0.2              pyhd8ed1ab_0    conda-forge
qdldl-python              0.1.5            py39h089d6f7_0    conda-forge
qt                        5.12.9               h126340a_4    conda-forge
qtawesome                 1.0.2              pyhd8ed1ab_0    conda-forge
qtconsole                 5.0.3              pyhd8ed1ab_0    conda-forge
qtpy                      1.9.0                      py_0    conda-forge
readline                  8.1                  h05e3726_0    conda-forge
regex                     2021.4.4         py39h89e85a6_0    conda-forge
requests                  2.25.1             pyhd3deb0d_0    conda-forge
rope                      0.18.0             pyh9f0ad1d_0    conda-forge
rtree                     0.9.7            py39h7d0d40a_1    conda-forge
scikit-learn              0.24.1           py39h1f17f7f_0    conda-forge
scipy                     1.6.2            py39h056f1c0_0    conda-forge
scotch                    6.0.9                h57311db_1    conda-forge
scs                       2.1.3            py39h5325327_0    conda-forge
send2trash                1.5.0                      py_0    conda-forge
setuptools                49.6.0           py39h6e9494a_3    conda-forge
six                       1.15.0             pyh9f0ad1d_0    conda-forge
slycot                    0.4.0.0          py39h9983580_4    conda-forge
sniffio                   1.2.0            py39h6e9494a_1    conda-forge
snowballstemmer           2.1.0              pyhd8ed1ab_0    conda-forge
sortedcontainers          2.3.0              pyhd8ed1ab_0    conda-forge
sphinx                    3.5.3              pyhd8ed1ab_0    conda-forge
sphinxcontrib-applehelp   1.0.2                      py_0    conda-forge
sphinxcontrib-devhelp     1.0.2                      py_0    conda-forge
sphinxcontrib-htmlhelp    1.0.3                      py_0    conda-forge
sphinxcontrib-jsmath      1.0.1                      py_0    conda-forge
sphinxcontrib-qthelp      1.0.3                      py_0    conda-forge
sphinxcontrib-serializinghtml 1.1.4                      py_0    conda-forge
spyder                    4.2.3            py39h6e9494a_0    conda-forge
spyder-kernels            1.10.2           py39h6e9494a_0    conda-forge
sqlite                    3.35.4               h44b9ce1_0    conda-forge
tbb                       2020.2               h940c156_4    conda-forge
tbb-devel                 2020.2               h940c156_4    conda-forge
terminado                 0.9.4            py39h6e9494a_0    conda-forge
testpath                  0.4.4                      py_0    conda-forge
textdistance              4.2.1              pyhd8ed1ab_0    conda-forge
threadpoolctl             2.1.0              pyh5ca1d4c_0    conda-forge
three-merge               0.1.1              pyh9f0ad1d_0    conda-forge
tk                        8.6.10               h0419947_1    conda-forge
toml                      0.10.2             pyhd8ed1ab_0    conda-forge
tornado                   6.1              py39hcbf5805_1    conda-forge
traitlets                 5.0.5                      py_0    conda-forge
typed-ast                 1.4.2            py39hcbf5805_0    conda-forge
typing_extensions         3.7.4.3                    py_0    conda-forge
tzdata                    2021a                he74cb21_0    conda-forge
ujson                     4.0.2            py39h9fcab8e_0    conda-forge
urllib3                   1.26.4             pyhd8ed1ab_0    conda-forge
utfcpp                    3.1.2                h694c41f_1    conda-forge
vtk                       9.0.1           no_osmesa_py39h49d52db_106    conda-forge
watchdog                  1.0.2            py39hcbf5805_1    conda-forge
wcwidth                   0.2.5              pyh9f0ad1d_2    conda-forge
webencodings              0.5.1                      py_1    conda-forge
wecopttool                0.0.1                     dev_0    <develop>
wheel                     0.36.2             pyhd3deb0d_0    conda-forge
wrapt                     1.12.1           py39hcbf5805_3    conda-forge
wurlitzer                 2.1.0            py39h6e9494a_0    conda-forge
x264                      1!161.3030           h35c211d_0    conda-forge
xarray                    0.17.0             pyhd8ed1ab_0    conda-forge
xz                        5.2.5                haf1e3a3_1    conda-forge
yaml                      0.2.5                haf1e3a3_0    conda-forge
yapf                      0.30.0             pyh9f0ad1d_0    conda-forge
zeromq                    4.3.4                h1c7c35f_0    conda-forge
zipp                      3.4.1              pyhd8ed1ab_0    conda-forge
zlib                      1.2.11            h7795811_1010    conda-forge
zstd                      1.4.9                h582d3a0_0    conda-forge
```

</p>
</details>

**Comment by mancellin** [comment]
@ryancoe That's a weird issue. I have no idea what could cause this.

The parallelism is just OpenMP: in the computation of the Green function, there is an OpenMP pragma at https://github.com/mancellin/capytaine/blob/87d09bf56dcd1d4a242a29805e9d49511b3947e1/capytaine/green_functions/Delhommeau_f90/matrices.f90#L114
Beside that, some matrix operations from Numpy and Scipy are also parallelized using OpenMP.

**Comment by ryancoe** [comment]
@mancellin @ssolson - I wanted to point you to this problem I noticed with WecOptTool, please take a look if you're able.

**Comment by mancellin** [comment]
Closing since #136 might solve the original issue. Feel free to reopen (or open a new issue) if you still have problems with the parallelization.
